sql/core/src/main/scala/org/apache/spark/sql/execution/CachedRDD.scala

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.{Dependency, NarrowDependency, Partition, TaskContext}
+import org.apache.spark.rdd.RDD
+
+/**
+ * The [[Partition]] used by [[CachedRDD]].
+ */
+case class CachedRDDPartition(
```

Comment:
can we add some PR comments to indicate which code are copied from somewhere and which code are new?

Reply:
Most are from `CoalescedRDD`, https://github.com/apache/spark/blob/5a56c17283103821714ffaaf1c764e05d0ff6b58/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala#L74-L78
that has similar use case to coalesce RDD partition. The main difference is `CachedRDD` know its target partitions which make things simple. All core methods in `CachedRDD ` can reuse the origin of previous RDD.

Reply:
So if that's the case, is there a reason why we can not implement this with `CoalescedRDD` and supplying the correct `partitionCoalescer: Option[PartitionCoalescer]`? It feels a little strange having an RDD classes code copied from core to SQL.